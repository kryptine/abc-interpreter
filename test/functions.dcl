definition module functions

f :: !Int -> a -> Int
const5 :: (.b -> Int)
const :: !.a .b -> .a
