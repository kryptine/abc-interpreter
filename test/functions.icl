module functions

import StdInt

Start = (square, sub5, sub5 0 10, sumints, reverse, foldr, ap1, ap3, map, repeat)

square :: Int -> Int
square x = x * x

sub5 :: Int Int Int Int Int -> Int
sub5 a b c d e = a - b - c - d - e

sumints :: [Int] -> Int
sumints []     = 0
sumints [x:xs] = x + sumints xs

reverse :: [a] -> [a]
reverse xs = rev [] xs
where
	rev new []     = new
	rev new [x:xs] = rev [x:new] xs

foldr :: (a b -> b) b [a] -> b
foldr op e []     = e
foldr op e [x:xs] = op x (foldr op e xs)

ap1 :: (Int -> Int) -> Int
ap1 f = f 42

ap3 :: (Int Int Int -> Int) -> Int
ap3 f = f 3 2 1

map :: (a -> b) [a] -> [b]
map f [x:xs] = [f x:map f xs]
map _ []     = []

repeat :: a -> [a]
repeat x = let xs = [x:xs] in xs
