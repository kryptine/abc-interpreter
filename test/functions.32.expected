(square.2,sub5.3,(sub5.3 0 10),sumints.4,reverse.5,foldr.6,ap1.7,ap3.8,map.9,repeat.10,([(IT1 37 '!'),(IT2 37 '!')],{InternalType2 (IT1 37 '!') 1,InternalType2 (IT2 37 '!') 2},[(InternalType2 (IT1 37 '!') 1),(InternalType2 (IT2 37 '!') 2)]))
halt at 54
63 524225 524288
